nilearn/decomposition/tests/test_decomposition_estimators.py

```diff
@@ -0,0 +1,312 @@
+        ):
+            est.fit(decomposition_data)
+    # but with surface images, the mask encompasses all vertices
+    # so it should have the same number of True vertices as the vertices
+    # in input images
+    elif data_type == "surface":
+        est.fit(decomposition_data)
+        assert est.masker_.n_elements_ == decomposition_data[0].mesh.n_vertices
+
+
+@pytest.mark.parametrize("estimator", [CanICA, _MultiPCA, DictLearning])
+@pytest.mark.parametrize("data_type", ["nifti", "surface"])
+def test_masker_attributes_with_fit(
+    data_type,
+    canica_data,
+    decomposition_mask_img,
+    decomposition_masker,
+    estimator,
+):
+    """Test mask_img_ properly set when passing mask_img or masker."""
+    # Passing mask_img
+    est = estimator(
+        n_components=3,
+        mask=decomposition_mask_img,
+        random_state=RANDOM_STATE,
+        smoothing_fwhm=None,
+    )
+    est.fit(canica_data)
+
+    check_decomposition_estimator(est, data_type)
+
+    # Passing masker
+    canica = estimator(
+        n_components=3,
+        mask=decomposition_masker,
+        random_state=RANDOM_STATE,
+        smoothing_fwhm=None,
+    )
+    canica.fit(canica_data)
+
+    check_decomposition_estimator(canica, data_type)
+
+
+@pytest.mark.parametrize("estimator", [CanICA, _MultiPCA, DictLearning])
+@pytest.mark.parametrize("data_type", ["nifti", "surface"])
+def test_pass_masker_arg_to_estimator(
+    data_type, affine_eye, decomposition_data_single_img, estimator
+):
+    """Masker arguments are passed to the estimator without fail."""
+    shape = (
+        decomposition_data_single_img.shape[:3]
+        if data_type == "nifti"
+        else (decomposition_data_single_img.mesh.n_vertices,)
+    )
+    est = estimator(
+        target_affine=affine_eye,
+        target_shape=shape,
+        n_components=3,
+        mask_strategy="background",
+        random_state=RANDOM_STATE,
+        smoothing_fwhm=None,
+    )
+
+    # for surface we should get a warning about target_affine, target_shape
+    # and mask_strategy being ignored
+    if data_type == "surface":
+        with pytest.warns(
+            UserWarning, match="The following parameters are not relevant"
+        ):
+            est.fit(decomposition_data_single_img)
+    elif data_type == "nifti":
+        est.fit(decomposition_data_single_img)
+
+    check_decomposition_estimator(est, data_type)
+
+
+@pytest.mark.timeout(0)
+@pytest.mark.parametrize("estimator", [CanICA, _MultiPCA, DictLearning])
+@pytest.mark.parametrize("data_type", ["nifti"])
+def test_nifti_maps_masker_(canica_data_single_img, estimator):
+    """Check deprecation of nifti_maps_masker_."""
+    est = estimator()
+
+    est.fit(canica_data_single_img)
+
+    with pytest.deprecated_call(
+        match="The 'nifti_maps_masker_' attribute is deprecated"
+    ):
+        est.nifti_maps_masker_  # noqa: B018
+
+
+# TODO passing confounds does not affect output with CanICA, DictLearning
```

Comment:
I'm a bit lost here. Why is it the case ?

Reply:
Not sure: I was just noting it down as something to investigate later (out of the scope of this PR), I was not trying to suggest that this is the expected behavior.

- I wonder if this relates to https://github.com/nilearn/nilearn/issues/3428